File: src/main/java/jp/co/cybird/barcodekanojoForGAM/BarcodeKanojoApp.java
package jp.co.cybird.barcodekanojoForGAM;

import android.app.Application;
import android.content.Intent;
import android.location.Location;
import android.location.LocationManager;
import android.os.Build;

import jp.co.cybird.barcodekanojoForGAM.billing.util.PurchaseApi;
import jp.co.cybird.barcodekanojoForGAM.core.BarcodeKanojo;
import jp.co.cybird.barcodekanojoForGAM.core.location.BestLocationListener;
import jp.co.cybird.barcodekanojoForGAM.core.model.User;

import com.goujer.barcodekanojo.core.util.DynamicImageCache;
import com.goujer.barcodekanojo.preferences.ApplicationSetting;

public class BarcodeKanojoApp extends Application {
    public static final String INTENT_ACTION_FULL_STORAGE = "jp.co.cybird.barcodekanojoForGAM.intent.action.FULL_STORAGE";
    public static final String INTENT_ACTION_LOGGED_OUT = "jp.co.cybird.barcodekanojoForGAM.intent.action.LOGGED_OUT";
    public static final String PACKAGE_NAME = "jp.co.cybird.barcodekanojoForGAM";
    private static final String TAG = "BarcodeKanojo";
    private BarcodeKanojo mBarcodeKanojo;
    private ApplicationSetting settings;
    private DynamicImageCache imageCache;
    private BestLocationListener mBestLocationListener = new BestLocationListener();
    private PurchaseApi mPurchaseApi;
    private String[] mUserGenderList;

	@Override
    public void onCreate() {
        super.onCreate();
        this.mBarcodeKanojo = new BarcodeKanojo();
        this.mBarcodeKanojo.setUser(new User());
        this.mPurchaseApi = new PurchaseApi(getApplicationContext());
        this.mUserGenderList = getResources().getStringArray(R.array.user_account_gender_list);

		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.FROYO) {
			System.setProperty("http.keepAlive", "false");
		}
		settings = new ApplicationSetting(this);
		mBarcodeKanojo.createHttpApi(settings.getServerHttps(), settings.getServerURL(), settings.getServerPort(), Defs.USER_AGENT(), Defs.USER_LANGUAGE());
		imageCache = new DynamicImageCache(settings.getCacheSize(), getBaseContext(), mBarcodeKanojo.getHttpApi());
    }

    public void changeLocate() {
        this.mBarcodeKanojo = new BarcodeKanojo();
        this.mUserGenderList = getResources().getStringArray(R.array.user_account_gender_list);
    }

    public String getUDID() {
        return "au_barcodekanojo";
    }

    public String getUUID() {
        return settings.getUUID();
    }

    public User getUser() {
        return this.mBarcodeKanojo.getUser();
    }

    public BarcodeKanojo getBarcodeKanojo() {
        return this.mBarcodeKanojo;
    }

    public PurchaseApi getPurchaseApi() {
        if (this.mPurchaseApi.getHelper() == null) {
            this.mPurchaseApi = new PurchaseApi(getApplicationContext());
        }
        return this.mPurchaseApi;
    }

	public ApplicationSetting getSettings() {
		return settings;
	}

    public DynamicImageCache getImageCache() {
		return imageCache;
	}

    public void logged_out() {
        sendBroadcast(new Intent(INTENT_ACTION_LOGGED_OUT));
    }

    public BestLocationListener requestLocationUpdates(boolean gps) {
        this.mBestLocationListener.register((LocationManager) getSystemService(LOCATION_SERVICE), gps);
        return this.mBestLocationListener;
    }

//    public BestLocationListener requestLocationUpdates(Observer observer) {
//        this.mBestLocationListener.addObserver(observer);
//        this.mBestLocationListener.register((LocationManager) getSystemService(LOCATION_SERVICE), true);
//        return this.mBestLocationListener;
//    }

    public void removeLocationUpdates() {
        this.mBestLocationListener.unregister((LocationManager) getSystemService(LOCATION_SERVICE));
    }

//    public void removeLocationUpdates(Observer observer) {
//        this.mBestLocationListener.deleteObserver(observer);
//        removeLocationUpdates();
//    }

	//TODO This is a quick fix, but a real in depth consensual location approximation should be used eventually.
    public Location getLastKnownLocation() {
        return null;
    	//return this.mBestLocationListener.getLastKnownLocation();
    }

//    public Location getLastKnownLocationOrThrow() {
//        Location location = this.mBestLocationListener.getLastKnownLocation();
//        if (location == null) {
//            return null;
//        }
//        return location;
//    }

//    public void clearLastKnownLocation() {
//        this.mBestLocationListener.clearLastKnownLocation();
//    }

//    public void setSavingListener(BaseDiskCache.BaseDiskCallBack listener) {
//        this.mSaveListener = listener;
//        loadResourceManagers();
//    }

//    public void removeSaveListener() {
//        this.mSaveListener = null;
//        loadResourceManagers();
//    }

    public String[] getUserGenderList() {
        this.mUserGenderList = getResources().getStringArray(R.array.user_account_gender_list);
        return this.mUserGenderList;
    }

    public void updateBCKApi(boolean useHttps, String newUrl, int port) {
		mBarcodeKanojo.createHttpApi(useHttps, newUrl, port, Defs.USER_AGENT(), Defs.USER_LANGUAGE());
	}

	public String getmApiBaseUrl() {
    	return mBarcodeKanojo.getmApiBaseUrl();
	}
}

File: app/src/main/java/com/goujer/barcodekanojo/BarcodeKanojoApp.kt
package com.goujer.barcodekanojo

import android.app.Application
import android.content.Intent
import android.location.Location
import android.location.LocationManager
import androidx.multidex.MultiDexApplication
import com.google.android.gms.security.ProviderInstaller
import com.goujer.barcodekanojo.core.BarcodeKanojo
import com.goujer.barcodekanojo.core.cache.DynamicImageCache
import com.goujer.barcodekanojo.core.model.User
import com.goujer.barcodekanojo.preferences.ApplicationSetting

import jp.co.cybird.barcodekanojoForGAM.core.location.BestLocationListener
import org.conscrypt.Conscrypt
import java.security.Security
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager


class BarcodeKanojoApp : MultiDexApplication() {
	lateinit var barcodeKanojo: BarcodeKanojo
		private set
	lateinit var imageCache: DynamicImageCache
		private set
	private val mBestLocationListener = BestLocationListener()

	private lateinit var mUserGenderList: Array<String>

	override fun onCreate() {
		super.onCreate()

		//Security Setup (Should allow newer TLS stuff)
		Security.insertProviderAt(Conscrypt.newProvider(), 1)

		initServerConnection()
		mUserGenderList = resources.getStringArray(R.array.user_account_gender_list)
		imageCache = DynamicImageCache((Runtime.getRuntime().maxMemory() / 1024L).toInt() / 6, baseContext)
	}

	@Deprecated("Deprecated in Java")
	override fun onLowMemory() {
		super.onLowMemory()
		imageCache.evictAll()
	}

	override fun onTrimMemory(level: Int) {
		super.onTrimMemory(level)
        when (level) {
            TRIM_MEMORY_RUNNING_MODERATE -> imageCache.trimToSize(imageCache.size()/2)
            TRIM_MEMORY_RUNNING_LOW -> imageCache.trimToSize(imageCache.size()/3)
            TRIM_MEMORY_RUNNING_CRITICAL -> imageCache.trimToSize(imageCache.size()/4)
            TRIM_MEMORY_UI_HIDDEN,
            TRIM_MEMORY_BACKGROUND,
            TRIM_MEMORY_MODERATE,
            TRIM_MEMORY_COMPLETE -> imageCache.memEvictAll()
        }
	}

	fun changeLocate() {
		barcodeKanojo = BarcodeKanojo(this)
		mUserGenderList = resources.getStringArray(R.array.user_account_gender_list)
	}

	val uDID: String
		get() = "au_barcodekanojo"

	val settings: ApplicationSetting
		get() = barcodeKanojo.settings

	val user: User?
		get() = barcodeKanojo.user

	fun initServerConnection() {
		barcodeKanojo = BarcodeKanojo(this)
		barcodeKanojo.user = User()
	}

	fun logged_out() {
		sendBroadcast(Intent(INTENT_ACTION_LOGGED_OUT))
	}

	fun requestLocationUpdates(gps: Boolean): BestLocationListener {
		mBestLocationListener.register(getSystemService(LOCATION_SERVICE) as LocationManager, gps)
		return mBestLocationListener
	}

	//    public BestLocationListener requestLocationUpdates(Observer observer) {
	//        this.mBestLocationListener.addObserver(observer);
	//        this.mBestLocationListener.register((LocationManager) getSystemService(LOCATION_SERVICE), true);
	//        return this.mBestLocationListener;
	//    }

	fun removeLocationUpdates() {
		mBestLocationListener.unregister(getSystemService(LOCATION_SERVICE) as LocationManager)
        //return this.mBestLocationListener.getLastKnownLocation();
	}

	//    public void removeLocationUpdates(Observer observer) {
	//        this.mBestLocationListener.deleteObserver(observer);
	//        removeLocationUpdates();
	//    }

	//TODO This is a quick fix, but a real in depth consensual location approximation should be used eventually.
	val lastKnownLocation: Location?
		get() = null

	//return this.mBestLocationListener.getLastKnownLocation();
	//    public Location getLastKnownLocationOrThrow() {
	//        Location location = this.mBestLocationListener.getLastKnownLocation();
	//        if (location == null) {
	//            return null;
	//        }
	//        return location;
	//    }

	//    public void clearLastKnownLocation() {
	//        this.mBestLocationListener.clearLastKnownLocation();
	//    }

	//    public void setSavingListener(BaseDiskCache.BaseDiskCallBack listener) {
	//        this.mSaveListener = listener;
	//        loadResourceManagers();
	//    }

	//    public void removeSaveListener() {
	//        this.mSaveListener = null;
	//        loadResourceManagers();
	//    }

	val userGenderList: Array<String>
		get() {
			mUserGenderList = resources.getStringArray(R.array.user_account_gender_list)
			return mUserGenderList
		}

	companion object {
		const val INTENT_ACTION_FULL_STORAGE = "jp.co.cybird.barcodekanojoForGAM.intent.action.FULL_STORAGE"
		const val INTENT_ACTION_LOGGED_OUT = "jp.co.cybird.barcodekanojoForGAM.intent.action.LOGGED_OUT"
	}
}